File: Java/46.java
import java.util.*;

class Solution46 {
    public static List<List<Integer>> permute(int[] nums) {
        // 思路: dfs
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        dfs(nums, res, ans, visited);
        return res;
    }

    public static void dfs(int[] nums, List<List<Integer>> res, List<Integer> ans, boolean[] visited){
        if (ans.size() == nums.length) {res.add(new ArrayList<Integer>(ans)); return;}
        for (int i = 0; i < nums.length; i++)
            if (!visited[i]){
                ans.add(nums[i]);
                visited[i] = true;
                dfs(nums, res, ans, visited);
                ans.remove(ans.size()-1);
                visited[i] = false;
            }
    }
}

File: Python/46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        def back(S):
            if len(S) == n:
                res.append(self.mycopy(S))
                return 
            for i in nums:
                if i not in S:
                    S.append(i)
                    back(S)
                    S.pop()
        back([])
        return res
    
    def mycopy(self, a):
        # 将列表a的值全部复制到b中，并且返回b
        b = a[:]
        return b